sklearn/mixture/gaussian_mixture.py

```diff
@@ -0,0 +1,673 @@
+        "full": _estimate_gaussian_covariance_full,
+        "tied": _estimate_gaussian_covariance_tied,
+        "diag": _estimate_gaussian_covariance_diag,
+        "spherical": _estimate_gaussian_covariance_spherical}
+
+    nk = resp.sum(axis=0) + 10 * np.finfo(float).eps
+    means = np.dot(resp.T, X) / nk[:, np.newaxis]
+    covariances = compute_covariance[covariance_type](
+        resp, X, nk, means, reg_covar)
+    nk /= X.shape[0]
+
+    return nk, means, covariances
+
+
+###############################################################################
+# Gaussian mixture probability estimators
+
+def _estimate_log_gaussian_prob_full(X, means, covariances):
+    """Estimate the log Gaussian probability for 'full' covariance.
+
+    Parameters
+    ----------
+    X : array-like, shape = (n_samples, n_features)
+
+    means : array-like, shape = (n_components, n_features)
+
+    covariances : array-like, shape = (n_components, n_features, n_features)
+
+    Returns
+    -------
+    log_prob : array, shape = (n_samples, n_components)
+    """
+    n_samples, n_features = X.shape
+    n_components = means.shape[0]
+    log_prob = np.empty((n_samples, n_components))
+    for k, (mu, cov) in enumerate(zip(means, covariances)):
+        try:
+            cov_chol = linalg.cholesky(cov, lower=True)
+        except linalg.LinAlgError:
+            raise ValueError("'covariances' must be symmetric, "
+                             "positive-definite")
+        cv_log_det = 2. * np.sum(np.log(np.diagonal(cov_chol)))
+        cv_sol = linalg.solve_triangular(cov_chol, (X - mu).T,
+                                         lower=True).T
+        log_prob[:, k] = - .5 * (n_features * np.log(2. * np.pi) +
+                                 cv_log_det +
+                                 np.sum(np.square(cv_sol), axis=1))
+    return log_prob
+
+
+def _estimate_log_gaussian_prob_tied(X, means, covariances):
+    """Estimate the log Gaussian probability for 'tied' covariance.
+
+    Parameters
+    ----------
+    X : array-like, shape = (n_samples, n_features)
+
+    means : array-like, shape = (n_components, n_features)
+
+    covariances : array-like, shape = (n_features, n_features)
+
+    Returns
+    -------
+    log_prob : array-like, shape = (n_samples, n_components)
+    """
+    n_samples, n_features = X.shape
+    n_components = means.shape[0]
+    log_prob = np.empty((n_samples, n_components))
+    try:
+        cov_chol = linalg.cholesky(covariances, lower=True)
+    except linalg.LinAlgError:
+        raise ValueError("'covariances' must be symmetric, positive-definite")
+    cv_log_det = 2. * np.sum(np.log(np.diagonal(cov_chol)))
+    for k, mu in enumerate(means):
+        cv_sol = linalg.solve_triangular(cov_chol, (X - mu).T,
+                                         lower=True).T
+        log_prob[:, k] = np.sum(np.square(cv_sol), axis=1)
+    log_prob = - .5 * (n_features * np.log(2. * np.pi) + cv_log_det + log_prob)
```

Comment:
I've try to find another way to do that. Unfortunately, this code is optimize for each covariance type.
The best I can propose is on that [GIST](https://gist.github.com/tguillemot/2374805a5251f08e630c#file-log_prob-py) but I don't like this solution, it increases a lot the code complexity.